File: analytics/src/test/java/io/appmetrica/analytics/impl/referrer/service/ReferrerManagerTest.java
package io.appmetrica.analytics.impl.referrer.service;

import io.appmetrica.analytics.coreutils.internal.logger.LoggerStorage;
import io.appmetrica.analytics.impl.referrer.common.ReferrerChosenListener;
import io.appmetrica.analytics.impl.referrer.common.ReferrerInfo;
import io.appmetrica.analytics.logger.appmetrica.internal.PublicLogger;
import io.appmetrica.analytics.testutils.CommonTest;
import io.appmetrica.analytics.testutils.MockedStaticRule;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatcher;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class ReferrerManagerTest extends CommonTest {

    @Mock
    private ReferrerHolder referrerHolder;
    private final String referrer = "Referrer from GP";
    private final ReferrerInfo.Source sourceGp = ReferrerInfo.Source.GP;
    private ReferrerInfo referrerInfo;
    @Mock
    private ReferrerChosenListener firstReferrerChosenListener;
    @Mock
    private ReferrerChosenListener secondReferrerChosenListener;
    private ReferrerManager referrerManager;

    @Mock
    private PublicLogger logger;

    @Rule
    public MockedStaticRule<LoggerStorage> loggerStorageMockedStatic = new MockedStaticRule<>(LoggerStorage.class);

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        when(LoggerStorage.getMainPublicOrAnonymousLogger()).thenReturn(logger);
        referrerInfo = new ReferrerInfo(referrer, 100500L, 200500L, sourceGp);
        referrerManager = new ReferrerManager(referrerHolder);
    }

    @Test
    public void subscribedToReferrerHolder() {
        ArgumentCaptor<ReferrerListenerNotifier> listenerCaptor = ArgumentCaptor.forClass(ReferrerListenerNotifier.class);
        verify(referrerHolder).subscribe(listenerCaptor.capture());
        ReferrerListenerNotifier listenerNotifier = listenerCaptor.getValue();
        assertThat(listenerNotifier).isInstanceOf(SimpleReferrerListenerNotifier.class);
        assertThat(listenerNotifier.getListener()).isSameAs(referrerManager);
    }

    @Test
    public void referrerIsRequested() {
        verify(referrerHolder).retrieveReferrerIfNeeded();
    }

    @Test
    public void handleReferrerThenAddTwoListeners() {
        referrerManager.handleReferrer(referrerInfo);
        referrerManager.addOneShotListener(firstReferrerChosenListener);
        referrerManager.addOneShotListener(secondReferrerChosenListener);
        verify(firstReferrerChosenListener).onReferrerChosen(referrerInfo);
        verify(secondReferrerChosenListener).onReferrerChosen(referrerInfo);
    }

    @Test
    public void handleNullReferrerThenAddTwoListeners() {
        referrerManager.handleReferrer(null);
        referrerManager.addOneShotListener(firstReferrerChosenListener);
        referrerManager.addOneShotListener(secondReferrerChosenListener);
        verify(firstReferrerChosenListener).onReferrerChosen(null);
        verify(secondReferrerChosenListener).onReferrerChosen(null);
    }

    @Test
    public void addTwoListenersThenHandleReferrer() {
        referrerManager.addOneShotListener(firstReferrerChosenListener);
        referrerManager.addOneShotListener(secondReferrerChosenListener);
        verifyNoMoreInteractions(firstReferrerChosenListener, secondReferrerChosenListener);
        referrerManager.handleReferrer(referrerInfo);
        verify(firstReferrerChosenListener).onReferrerChosen(referrerInfo);
        verify(secondReferrerChosenListener).onReferrerChosen(referrerInfo);
    }

    @Test
    public void addTwoListenersThenHandleNullReferrer() {
        referrerManager.addOneShotListener(firstReferrerChosenListener);
        referrerManager.addOneShotListener(secondReferrerChosenListener);
        verifyNoMoreInteractions(firstReferrerChosenListener, secondReferrerChosenListener);
        referrerManager.handleReferrer(null);
        verify(firstReferrerChosenListener).onReferrerChosen(null);
        verify(secondReferrerChosenListener).onReferrerChosen(null);
    }

    @Test
    public void addOneListenerHandleReferrerAndAddAnother() {
        referrerManager.addOneShotListener(firstReferrerChosenListener);
        verifyNoMoreInteractions(firstReferrerChosenListener);
        referrerManager.handleReferrer(referrerInfo);
        verify(firstReferrerChosenListener).onReferrerChosen(referrerInfo);
        referrerManager.addOneShotListener(secondReferrerChosenListener);
        verify(secondReferrerChosenListener).onReferrerChosen(referrerInfo);
    }

    @Test
    public void handleReferrerTwice() {
        referrerManager.addOneShotListener(firstReferrerChosenListener);
        verifyNoMoreInteractions(firstReferrerChosenListener);
        referrerManager.handleReferrer(referrerInfo);
        referrerManager.handleReferrer(referrerInfo);
        verify(firstReferrerChosenListener, times(1)).onReferrerChosen(referrerInfo);
    }

    @Test
    public void handleReferrerTwiceFirstIsNull() {
        referrerManager.addOneShotListener(firstReferrerChosenListener);
        verifyNoMoreInteractions(firstReferrerChosenListener);
        referrerManager.handleReferrer(null);
        referrerManager.handleReferrer(referrerInfo);
        verify(firstReferrerChosenListener, times(1)).onReferrerChosen(null);
        verify(firstReferrerChosenListener, never()).onReferrerChosen(referrerInfo);
    }

    @Test
    public void handleReferrerWritePublicLog() {
        referrerManager.handleReferrer(referrerInfo);
        verify(logger).info(
            argThat(new ArgumentMatcher<String>() {
                @Override
                public boolean matches(String argument) {
                    return argument.contains("Received referrer from source");
                }
            }),
            eq(sourceGp.value),
            eq(referrer)
        );
    }

    @Test
    public void handleReferrerWritePublicLogIfReferrerIsNull() {
        referrerManager.handleReferrer(null);
        verifyNoInteractions(logger);
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/referrer/service/ReferrerManagerTest.kt
package io.appmetrica.analytics.impl.referrer.service

import android.content.Context
import io.appmetrica.analytics.impl.referrer.service.listener.SafeReferrerListener
import io.appmetrica.analytics.impl.referrer.service.provider.ReferrerProvider
import io.appmetrica.analytics.impl.referrer.service.provider.ReferrerProviderFactory
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.constructionRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.mockito.kotlin.any
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.doThrow
import org.mockito.kotlin.mock
import org.mockito.kotlin.times
import org.mockito.kotlin.verify

internal class ReferrerManagerTest : CommonTest() {
    private val context: Context = mock()
    private val referrerProvider: ReferrerProvider = mock()
    private val listenerCaptor = argumentCaptor<ReferrerListener>()
    private val result = mock<ReferrerResult>()
    private val externalListener: ReferrerListener = mock()

    @get:Rule
    val referrerProviderFactoryRule = constructionRule<ReferrerProviderFactory> {
        on { create(context) } doReturn referrerProvider
    }

    private val referrerManager by setUp { ReferrerManager(context) }

    @Test
    fun `constructor with Context calls factory with correct arguments`() {
        ReferrerManager(context)

        val factories = referrerProviderFactoryRule.constructionMock.constructed()
        verify(factories[0]).create(context)
    }

    @Test
    fun `getCachedReferrer returns null initially`() {
        assertThat(referrerManager.getCachedReferrer()).isNull()
    }

    @Test
    fun `getCachedReferrer returns cached result after warmUpReferrer`() {
        referrerManager.warmUpReferrer()

        verify(referrerProvider).requestReferrer(listenerCaptor.capture())

        listenerCaptor.firstValue.onResult(result)
        assertThat(referrerManager.getCachedReferrer()).isEqualTo(result)
    }

    @Test
    fun `requestReferrer forwards listener to provider`() {
        referrerManager.requestReferrer(externalListener)

        verify(referrerProvider).requestReferrer(listenerCaptor.capture())

        listenerCaptor.firstValue.onResult(result) // first listener from init

        verify(externalListener).onResult(result)
    }

    @Test
    fun `requestReferrer wraps listener in SafeReferrerListener`() {
        referrerManager.requestReferrer(externalListener)

        verify(referrerProvider).requestReferrer(listenerCaptor.capture())

        assertThat(listenerCaptor.firstValue).isInstanceOf(SafeReferrerListener::class.java)
    }

    @Test
    fun `requestReferrer does not crash if listener throws exception`() {
        val throwingListener: ReferrerListener = mock {
            on { onResult(any()) } doThrow RuntimeException("Test exception")
        }

        referrerManager.requestReferrer(throwingListener)

        verify(referrerProvider).requestReferrer(listenerCaptor.capture())

        // Should not throw
        listenerCaptor.firstValue.onResult(result)
    }

    @Test
    fun `multiple requestReferrer calls do not affect cache`() {
        referrerManager.warmUpReferrer()

        val initListenerCaptor = argumentCaptor<ReferrerListener>()
        verify(referrerProvider).requestReferrer(initListenerCaptor.capture())

        val initResult = mock<ReferrerResult>()
        initListenerCaptor.firstValue.onResult(initResult)
        assertThat(referrerManager.getCachedReferrer()).isEqualTo(initResult)

        val externalListener1: ReferrerListener = mock()
        val externalListener2: ReferrerListener = mock()

        referrerManager.requestReferrer(externalListener1)
        referrerManager.requestReferrer(externalListener2)

        val allListenerCaptor = argumentCaptor<ReferrerListener>()
        verify(referrerProvider, times(3)).requestReferrer(allListenerCaptor.capture())

        val externalResult1 = mock<ReferrerResult>()
        val externalResult2 = mock<ReferrerResult>()

        allListenerCaptor.secondValue.onResult(externalResult1)
        allListenerCaptor.thirdValue.onResult(externalResult2)

        // Cache should still be the initial result, not affected by external listeners
        assertThat(referrerManager.getCachedReferrer()).isEqualTo(initResult)
    }

    @Test
    fun `external listener receives result even if another listener throws exception`() {
        val throwingListener: ReferrerListener = mock {
            on { onResult(any()) } doThrow RuntimeException("Test exception")
        }
        val normalListener: ReferrerListener = mock()

        referrerManager.requestReferrer(throwingListener)
        referrerManager.requestReferrer(normalListener)

        verify(referrerProvider, times(2)).requestReferrer(listenerCaptor.capture())

        // Both wrapped listeners should be called, no exception should propagate
        listenerCaptor.firstValue.onResult(result)
        listenerCaptor.secondValue.onResult(result)

        verify(normalListener).onResult(result)
    }
}
